File: problems/3/solution.py
# Accepted: 45ms (96.20%), 16.71MB (57.29%)

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        right = 0
        lastOccur = dict()
        longest = 0

        while right < len(s):
            ch = s[right]

            if ch not in lastOccur or lastOccur[ch] < left: # no repeating character contained
                longest = max(longest, right - left + 1)
            else:
                left = lastOccur[ch]+1

            lastOccur[ch] = right
            right += 1

        return longest

File: problems/3/solution.ts
// Accepted: 115ms (35.87%), 57.48MB (10.88%)

function lengthOfLongestSubstring(s: string): number {
    let left = 0;
    let right = 0;
    const lastOccur = new Map<number, number>();
    let longest = 0;

    while (right < s.length) {
        const ch = s[right];

        if (lastOccur[ch] === undefined || lastOccur[ch] < left) { // no repeating character contained
            longest = Math.max(longest, right-left+1);
        } else {
            left = lastOccur[ch]+1;
        }

        lastOccur[ch] = right;
        ++right;
    }

    return longest;
};
